qiskit/extensions/quantum_initializer/state_preparation.py

```diff
@@ -0,0 +1,462 @@
+            # work out which rotations must be done to disentangle the LSB
+            # qubit (we peel away one qubit at a time)
+            (remaining_param, thetas, phis) = StatePreparation._rotations_to_disentangle(
+                remaining_param
+            )
+
+            # perform the required rotations to decouple the LSB qubit (so that
+            # it can be "factored" out, leaving a shorter amplitude vector to peel away)
+
+            add_last_cnot = True
+            if np.linalg.norm(phis) != 0 and np.linalg.norm(thetas) != 0:
+                add_last_cnot = False
+
+            if np.linalg.norm(phis) != 0:
+                rz_mult = self._multiplex(RZGate, phis, last_cnot=add_last_cnot)
+                circuit.append(rz_mult.to_instruction(), q[i : self.num_qubits])
+
+            if np.linalg.norm(thetas) != 0:
+                ry_mult = self._multiplex(RYGate, thetas, last_cnot=add_last_cnot)
+                circuit.append(ry_mult.to_instruction().reverse_ops(), q[i : self.num_qubits])
+        circuit.global_phase -= np.angle(sum(remaining_param))
+        return circuit
+
+    @staticmethod
+    def _rotations_to_disentangle(local_param):
+        """
+        Static internal method to work out Ry and Rz rotation angles used
+        to disentangle the LSB qubit.
+        These rotations make up the block diagonal matrix U (i.e. multiplexor)
+        that disentangles the LSB.
+
+        [[Ry(theta_1).Rz(phi_1)  0   .   .   0],
+        [0         Ry(theta_2).Rz(phi_2) .  0],
+                                    .
+                                        .
+        0         0           Ry(theta_2^n).Rz(phi_2^n)]]
+        """
+        remaining_vector = []
+        thetas = []
+        phis = []
+
+        param_len = len(local_param)
+
+        for i in range(param_len // 2):
+            # Ry and Rz rotations to move bloch vector from 0 to "imaginary"
+            # qubit
+            # (imagine a qubit state signified by the amplitudes at index 2*i
+            # and 2*(i+1), corresponding to the select qubits of the
+            # multiplexor being in state |i>)
+            (remains, add_theta, add_phi) = StatePreparation._bloch_angles(
+                local_param[2 * i : 2 * (i + 1)]
+            )
+
+            remaining_vector.append(remains)
+
+            # rotations for all imaginary qubits of the full vector
+            # to move from where it is to zero, hence the negative sign
+            thetas.append(-add_theta)
+            phis.append(-add_phi)
+
+        return remaining_vector, thetas, phis
+
+    @staticmethod
+    def _bloch_angles(pair_of_complex):
+        """
+        Static internal method to work out rotation to create the passed-in
+        qubit from the zero vector.
+        """
+        [a_complex, b_complex] = pair_of_complex
+        # Force a and b to be complex, as otherwise numpy.angle might fail.
+        a_complex = complex(a_complex)
+        b_complex = complex(b_complex)
```

Comment:
I think this comment is probably out-of-date - `np.angle(1)` will work just fine, for example - but really if this is necessary (say if they might be bound `ParameterExpression`) then it might have been better to handle this at the parameter validation stage.